src/source/Response.c

```diff
@@ -93,12 +97,29 @@ STATUS freeCurlResponse(PCurlResponse* ppCurlResponse)
         pCallbacksProvider = pCurlResponse->pCurlRequest->pCurlApiCallbacks->pCallbacksProvider;
     }
 
-    if (pCallbacksProvider != NULL && IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
-        locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
-        // if it fail to acquire the lock that means another thread is tearing down the response already.
-        // thus current thread can return early.
-        // TODO: eliminate the possibility of two threads calling freeCurlResponse.
-        CHK(locked, retStatus);
+    if (pCallbacksProvider != NULL) {
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
+            locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
+            // if it fail to acquire the lock that means another thread is tearing down the response already.
+            // thus current thread can return early.
+            // TODO: eliminate the possibility of two threads calling freeCurlResponse.
+            CHK(locked, retStatus);
+        }
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->dataAvailableLock)) {
+            dataAvailableLocked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->dataAvailableLock);

```

Comment:
You should not lock the dataAvailableLock and signal as if there is a CVAR awaiting, it won't be able to re-acquire the lock to "wake up".

Reply:
Ok I see the reason it won't is because we will have deleted the mutex before the other thread we signaled gets a chance to acquire it, so it cannot acquire it.